### PR TITLE
Using Imbalanced-Learn to handle imbalanced text data in Natural Language Processing

### LINKED ISSUE #7185: [Machine Learning] Using Imbalanced-Learn to handle imbalanced text data in natural language processing
#### Proposed title of article
[Machine Learning] Using Imbalanced-Learn to handle imbalanced text data in natural language processing

#### Proposed article introduction
Imbalanced-Learn is a Python module that helps in balancing the datasets which are highly skewed or biased towards some classes. Thus, it helps in resampling the classes which are otherwise oversampled or under-sampled. It offers a number of re-sampling techniques commonly used in datasets showing strong between-class imbalance.

Most models that are built using the imbalanced data will have a bias towards predicting the majority classes in the dataset and ignore the minority classes. Class imbalance is identified by observing the distribution of classes within your data or using visualizing libraries to plot the class distributions.

Imbalanced-Learn has various techniques that will be used to handle this problem. These techniques are such as Random Under sampling technique resample all classes but the majority class, Random Oversampling technique: resample all classes but the majority class.
In this tutorial, we will discuss these techniques in natural language processing and implement them practically using Python. We will also perform all the text preprocessing steps.

#### Key takeaways 
1. Text preprocessing (we will perform word lemmatization, removing stop words, and removing punctuations).
2.  Performing data augmentation using NLP Transformers.
3. Installing Imbalanced-Learn.
4. Exploring Imbalanced-Learn methods and techniques.
5. Implementing `RandomOverSampler` to balance the classes.

#### Article quality
This tutorial is unique because we will implement text preprocessing in detail so that a reader can understand the concepts in natural language processing. We will create a custom function for word lemmatization, removing stop words, removing punctuations, removing extra white spaces, and removing numbers. This ensures we have a clean dataset for natural language processing.  We will also perform data augmentation using NLP Transformers. Finally, we will implement `RandomOverSampler` to balance the classes.
#### References
Please list links to any published content/research that you intend to use to support/guide this article.

### Conclusion
Finally, remove the Pre-Submission advice section and all our blockquoted notes as you fill in the form before you submit.
We look forwarding to reviewing your topic suggestion.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by donnexwafula33**
## Our **Peer Reviewer and Content Moderator teams do NOT provide any revisions services.** All revisions and edits should be completed on your own forked repo (as to not take up room in the queue) in order for our team to review them all in a timely manner. 

**ANY ARTICLE SUBMITTED WITH GLARING ERRORS WILL BE IMMEDIATELY CLOSED.**

>As a rule of thumb - please be sure to only submit articles (pull requests) that are fully polished and ready to be published. Be sure to go through our resources documents for extra and 3rd party (vetted) resources to help improve overall technical writing.

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #7185

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by ahmadmardeni1** [comment]
Plagiarism check done ✅

**Comment by wilkistermumbi** [comment]
@donnexwafula33 Please pass your article through Grammarly. Preferrably, Grammarly premium to correct grammatical errors.

**Comment by donnexwafula33** [comment]
I have corrected the grammatical errors using Grammarly premium suggestions. @wilkistermumbi

File: content/articles/imbalanced-learn-to-handle-imbalanced-text-data-in-natural-language-processing/index.md [hunk 1]
```diff
@@ -0,0 +1,638 @@
...
+- [Conclusion](#conclusion)
+- [References](#references)
+
+### Prerequisites
+Ensure you know the following concepts to follow along easily and understand the tutorial:
+
+- Know [Python Programming] (https://www.w3schools.com/python/) concepts.
+- Know natural language concepts and how to [implemment an NLP model](/engineering-education/building-autocorrect-feature-using-nlp-with-python/).
+- Understand [machine learning modelling] (/engineering-education/house-price-prediction/) and other related concepts.
+- Implement the spam classification model in [Google Colab](https://research.google.com/colaboratory/) to get the same results.
```
Comment (wilkistermumbi) [inline content/articles/imbalanced-learn-to-handle-imbalanced-text-data-in-natural-language-processing/index.md:56]:
```suggestion
To follow along easily and understand the tutorial, ensure you know the following concepts:
- [Python Programming](https://www.w3schools.com/python/) concepts.
- Natural language concepts and how to [implement an NLP model](/engineering-education/building-autocorrect-feature-using-nlp-with-python/).
- Understand [machine learning modelling](/engineering-education/house-price-prediction/) and other related concepts.
- Implement the spam classification model in [Google Colab](https://research.google.com/colaboratory/) to get the same results.
```

**Comment by wilkistermumbi** [comment]
Please shorten this URL `/imbalanced-learn-to-handle-imbalanced-text-data-in-natural-language-processing/`

It is too long. Always keep your URLs short.

File: content/articles/imbalanced-learn-to-handle-imbalanced-text-data-in-natural-language-processing/index.md [hunk 2]
```diff
@@ -0,0 +1,638 @@
...
+
+### Table of contents
+- [Prerequisites](#prerequisites)
+- [Spam classification dataset](#spam-classification-dataset)
+- [Renaming the dataset column](#renaming-the-dataset-column)
+- [Calculating the length of each data sample](#calculating-the-length-of-each-data-sample)
+- [Text cleaning](#text-cleaning)
+- [Installing NLTK](#installing-nltk)
+- [Creating the custom functions](#creating-the-custom-functions)
+- [Converting the class labels into integer values](#converting-the-class-labels-into-interger-values)
```
Comment (wilkistermumbi) [inline content/articles/imbalanced-learn-to-handle-imbalanced-text-data-in-natural-language-processing/index.md:35]: Fix this heading. It is not rendering.

File: content/articles/imbalanced-learn-to-handle-imbalanced-text-data-in-natural-language-processing/index.md [hunk 3]
```diff
@@ -0,0 +1,638 @@
...
+- Know natural language concepts and how to [implemment an NLP model](/engineering-education/building-autocorrect-feature-using-nlp-with-python/).
+- Understand [machine learning modelling] (/engineering-education/house-price-prediction/) and other related concepts.
+- Implement the spam classification model in [Google Colab](https://research.google.com/colaboratory/) to get the same results.
+
+### Spam classification dataset
+We will use the SMS collection dataset to train the NLP model. It has two labeled classes. (`spam` and `ham`). 
+
+The `spam` class contains all the spam SMS. The `ham` class has all the SMS that are not spam. 
+
+The NLP model will classify an SMS as either spam or not spam. Ensure you can download the SMS collection dataset from [here](https://drive.google.com/file/d/1OLdz-fSUBEARnMqUdrTW4bkmGy8yUpCE/view?usp=sharing).
```
Comment (wilkistermumbi) [inline content/articles/imbalanced-learn-to-handle-imbalanced-text-data-in-natural-language-processing/index.md:63]: As a rule of thumb, ensure that your paragraphs are at least three lines long before adding spaces. Fix this issue throughout your tutorial

File: content/articles/imbalanced-learn-to-handle-imbalanced-text-data-in-natural-language-processing/index.md [hunk 4]
```diff
@@ -0,0 +1,638 @@
...
+df['length'] = df['text'].apply(lambda x: len(x))
+```
+To see the contents of the new `length` column, input this code:
+
+```python
+df.head()
+```
+It gives this output:
+
+![New length column](/engineering-education/imbalanced-learn-to-handle-imbalanced-text-data-in-natural-language-processing/new-length-column.png). We will start cleaning the text.
```
Comment (wilkistermumbi) [inline content/articles/imbalanced-learn-to-handle-imbalanced-text-data-in-natural-language-processing/index.md:148]:
```suggestion
![New length column](/engineering-education/imbalanced-learn-to-handle-imbalanced-text-data-in-natural-language-processing/new-length-column.png). 

We will begin cleaning the text.
```

File: content/articles/imbalanced-learn-to-handle-imbalanced-text-data-in-natural-language-processing/index.md [hunk 5]
```diff
@@ -0,0 +1,638 @@
...
+
+### Text cleaning
+Before building the spam classification model, we will clean the dataset to have the required format. Many text cleaning steps will format the text. Some of them are, removing unnecessary words, punctuation, stop words, white spaces, and unnecessary symbols from the text dataset.
```
Comment (wilkistermumbi) [inline content/articles/imbalanced-learn-to-handle-imbalanced-text-data-in-natural-language-processing/index.md:151]:
```suggestion
Before building the spam classification model, we will clean the dataset to have the required format. Many text cleaning steps will format the text. It includes removing unnecessary words, punctuation, stop words, white spaces, and unnecessary symbols from the text dataset.
```

File: content/authors/donnex-wafula/index.md [hunk 6]
```diff
@@ -0,0 +1,8 @@
+---
+title: Donnex Wafula
+type: authors
+github: https://github.com/donnexwafula33
+images:
+  - url: /engineering-education/authors/donnex-wafula/avatar.jpg 
+---
+Donnex Wafula is an undergraduate Computer Science student. His interests are web development and machine learning. Donnex has a great passion for developing web applications, technical writing and building machine learning models. He specializes in natural language processing and image classification.
```
Comment (wilkistermumbi) [inline content/authors/donnex-wafula/index.md:8]: Please use the format shown in this [documentation](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/UPLOAD_INSTRUCTIONS.md) to fix your author bio

File: content/articles/imbalanced-learn-to-handle-imbalanced-text-data-in-natural-language-processing/index.md [hunk 7]
```diff
@@ -0,0 +1,638 @@
...
+print(X_tf)
+```
+It gives this output:
+
+![Array](/engineering-education/imbalanced-learn-to-handle-imbalanced-text-data-in-natural-language-processing/array.png)
+
+We will feed the array into the model for training. 
+
+### Splitting the vectorized dataset
+We will split the vectorized dataset into two portions/sets. The first portion will be for model training and the
```
Comment (wilkistermumbi) [inline content/articles/imbalanced-learn-to-handle-imbalanced-text-data-in-natural-language-processing/index.md:421]: This sentence seems to be incomplete.

File: content/articles/imbalanced-learn-to-handle-imbalanced-text-data-in-natural-language-processing/index.md [hunk 8]
```diff
@@ -0,0 +1,638 @@
...
+This accuracy score is `87.623%`. It shows the ratio of the accurately predicted data samples to the total data samples in the testing set. 
+
+We have built the model without class balancing. The next step is to implement the same model but use Imbalanced-learn to balance the classes.
+
+### Implementing Imbalanced-learn
+We install the library as follows:
+
+```python
+!pip install imbalanced-learn
+```
```
Comment (wilkistermumbi) [inline content/articles/imbalanced-learn-to-handle-imbalanced-text-data-in-natural-language-processing/index.md:492]: This is a `bash` command

**Comment by wilkistermumbi** [comment]
@donnexwafula33 This a very informative and educative tutorial. Please work on the revisions before we can move on for final review.

**Comment by donnexwafula33** [comment]
I have worked on the revisions @wilkistermumbi

**Comment by hectorkambow** [comment]
Jumping on this one

**Pull request merged**